File: js/idb-filestore.js
window.indexedDB = window.indexedDB || window.mozIndexedDB || window.webkitIndexedDB || window.msIndexedDB;
window.IDBTransaction = window.IDBTransaction || window.webkitIDBTransaction || window.msIDBTransaction || {READ_WRITE: "readwrite"};
window.IDBKeyRange = window.IDBKeyRange || window.webkitIDBKeyRange || window.msIDBKeyRange;

class IDBFileStore {
    #dbName;
    #storeName;
    #version;
    #request;
    #db;
    constructor(dbName = "Cache", storeName = "Cache"){
        this.#dbName = dbName;
        this.#storeName = storeName;
        this.#version = 1; //manually increment when update the store
    }

    open(onsuccess, onerror){
        if(!onsuccess){
            throw "Missing onsuccess handler";
        }
        if(!onerror){
            throw "Missing onerror handler";
        }
        if(!this.#dbName || !this.#storeName){
            throw "Missing name(s)";
        }
        if(this.#db){
            throw "Already open!";
        }
        var store = this;
        let request = window.indexedDB.open(this.#dbName, this.#version);
        request.onerror = onerror;
        request.onupgradeneeded = function(event) {
          var db = event.target.result;
          if(event.oldVersion < 1){ // new DB
              var objectStore = db.createObjectStore(store.#storeName, {keyPath: ["storeId", "fileId"]});
          }
        };
        request.onsuccess = function(event){
            store.#db = event.target.result;
            onsuccess();
        };
    }

    #validate(storeId, fileId){
        if(!this.#db){
            throw "Not open!";
        }
        if(typeof fileId !== 'number'){
            throw "fileId must be a number!";
        }
        if(typeof storeId !== 'number'){
           throw "storeId must be a number!";
        }
    }

    write(storeId, fileId, data, onsuccess, onerror){
        this.#validate(storeId, fileId);
        if(!(data instanceof Int8Array)) {
            throw "data must be an Int8Array!";
        }
        var filestore = this;

        let store = filestore.#db.transaction(filestore.#storeName, "readwrite").objectStore(filestore.#storeName);
        let request = store.put({storeId:storeId, fileId: fileId, data: data});
        request.onerror = onerror;
        request.onsuccess = function(event){
            onsuccess(request.result);
        }
    }

    read(storeId, fileId, onsuccess, onerror) {
        this.#validate(storeId, fileId);
        var filestore = this;
        let store = filestore.#db.transaction(filestore.#storeName, "readonly").objectStore(filestore.#storeName);
        let request = store.get([storeId, fileId]);
        request.onerror = onerror;
        request.onsuccess = function(event){
            if(request.result){
                onsuccess(request.result.data);
            }else{
                onsuccess(null);
            }
        };
    }

    static isSupported(){
        return typeof window.indexedDB !== 'undefined';
    }
}

File: src/main/java/web/impl/js/IDBFileStore.java
package web.impl.js;

import org.teavm.jso.JSBody;
import org.teavm.jso.JSFunctor;
import org.teavm.jso.JSObject;
import org.teavm.jso.typedarrays.Int8Array;

public abstract class IDBFileStore implements JSObject {

    @JSBody(script = "return typeof IDBFileStore !== 'undefined'")
    static native boolean exists();

    @JSBody(params = {"dbName", "storeName"}, script = "return new IDBFileStore(dbName, storeName)")
    static native IDBFileStore create(String dbName, String storeName);

    public abstract void open(NoArgCallback onsuccess, OneArgCallback<JSObject> onerror);

    public abstract void read(int storeId, int fileId, OneArgCallback<Int8Array> onsuccess, OneArgCallback<JSObject> onerror);

    public abstract void write(int storeId, int fileId, Int8Array data, NoArgCallback onsuccess, OneArgCallback<JSObject> onerror);

    @JSFunctor
    public interface  NoArgCallback extends JSObject{
        void apply();
    }

    @JSFunctor
    public interface OneArgCallback<T extends JSObject> extends JSObject{
        void apply(T data);
    }
}